### PR TITLE
feat(raftwal): Add support for encryption in raftwal

### PR DESCRIPTION
**Pull request by ahsanbarkati**
This PR adds support for encryption of the data in raftwal. It also adds some of the related tests.

This change is [Reviewable](https://reviewable.io/reviews/dgraph-io/dgraph/6714)

 
 
 

Docs Preview: [Dgraph Preview](https://dgraph-df71811c01-102020.surge.sh)

### CODE REVIEW

File: raftwal/log.go [hunk 1]
```diff
@@ -43,12 +48,16 @@ const (
 	maxNumEntries = 30000
 	// logFileOffset is offset in the log file where data is stored.
 	logFileOffset = 1 << 20 // 1MB
+	// encryptionKeyOffset is offset in the log file where encryption key is stored.
+	encryptionKeyOffset = logFileOffset - 16 // 1MB - 16B
```
Comment (jarifibrahim) [inline raftwal/log.go:52]: Mention what 16 represents.

File: raftwal/log.go [hunk 2]
```diff
@@ -242,3 +295,41 @@ func getLogFiles(dir string) ([]*logFile, error) {
...
+func (lf *logFile) bootstrap() error {
+	// registry is nil if we don't have encryption enabled.
+	if lf.registry == nil {
+		return nil
+	}
+
+	var err error
+	// generate data key for the log file.
+	if lf.dataKey, err = lf.registry.LatestDataKey(); err != nil {
+		return y.Wrapf(err, "Error while retrieving datakey in logFile.bootstarp")
```
Comment (jarifibrahim) [inline raftwal/log.go:322]: bootstarp => bootstrap

File: raftwal/log.go [hunk 3]
```diff
@@ -84,21 +97,52 @@ func logFname(dir string, id int64) string {
...
-		fid:      fid,
+			keyID := binary.BigEndian.Uint64(buf[:8])
+			var dk *pb.DataKey
+			// retrieve datakey from the keyID of the logfile.
+			if dk, err = lf.registry.DataKey(keyID); err != nil {
+				return nil, err
+			}
+			lf.dataKey = dk
+			lf.baseIV = buf[8:]
+			y.AssertTrue(len(lf.baseIV) == 8)
```
Comment (jarifibrahim) [inline raftwal/log.go:144]: baseIVSize

File: raftwal/log.go [hunk 4]
```diff
@@ -84,21 +97,52 @@ func logFname(dir string, id int64) string {
...
-		MmapFile: mf,
-		fid:      fid,
+			keyID := binary.BigEndian.Uint64(buf[:8])
+			var dk *pb.DataKey
+			// retrieve datakey from the keyID of the logfile.
+			if dk, err = lf.registry.DataKey(keyID); err != nil {
+				return nil, err
+			}
+			lf.dataKey = dk
+			lf.baseIV = buf[8:]
```
Comment (jarifibrahim) [inline raftwal/log.go:143]: baseIVSize

File: raftwal/log.go [hunk 5]
```diff
@@ -129,6 +173,15 @@ func (lf *logFile) GetRaftEntry(idx int) raftpb.Entry {
 			re.Data = append(re.Data, data...)
 		}
 	}
+	// Decrypt the data if encryption in enabled.
```
Comment (jarifibrahim) [inline raftwal/log.go:176]: in => is

File: raftwal/wal.go [hunk 6]
```diff
@@ -166,10 +170,24 @@ func (l *wal) AddEntries(entries []raftpb.Entry) error {
...
+			curr := l.current
+			if err := y.XORBlockStream(
+				&ebuf, re.Data, curr.dataKey.Data, curr.generateIV(uint64(offset))); err != nil {
+				return err
+			}
+			// Allocate slice for the encoded data and copy encoded bytes.
+			destBuf, next = l.current.AllocateSlice(len(ebuf.Bytes()), offset)
+			x.AssertTrue(copy(destBuf, ebuf.Bytes()) == len(ebuf.Bytes()))
+		} else {
+			destBuf, next = l.current.AllocateSlice(len(re.Data), offset)
```
Comment (jarifibrahim) [inline raftwal/wal.go:188]: This can be moved out of the else.

File: raftwal/wal.go [hunk 7]
```diff
@@ -166,10 +170,24 @@ func (l *wal) AddEntries(entries []raftpb.Entry) error {
...
+
+		// If encryption is enabled then encrypt the data.
+		if x.WorkerConfig.EncryptionKey != nil {
+			var ebuf bytes.Buffer
+			curr := l.current
+			if err := y.XORBlockStream(
+				&ebuf, re.Data, curr.dataKey.Data, curr.generateIV(uint64(offset))); err != nil {
+				return err
+			}
+			// Allocate slice for the encoded data and copy encoded bytes.
```
Comment (jarifibrahim) [inline raftwal/wal.go:184]: re.Data = ebuf.Bytes()

**Review by jarifibrahim — APPROVED** [review]
Got some comments.

**Review by manishrjain — CHANGES_REQUESTED** [review]
Once you write the key ID to the file, use that to determine if you should encrypt any data written to that file or not. Don't continuously check for the WorkerConfig. Do the same when reading data, see if it has a key ID, if so, look into the key registry. If we have a valid key ID, but no corresponding key registry, then we should error out.

> Dismissed @codelingo[bot] and @jarifibrahim from 2 discussions.
> *[Reviewable](https://reviewable.io/reviews/dgraph-io/dgraph/6714#-:-MJX3S0U8gdrM2tBmZCd:b-lv771a)* status: 0 of 6 files reviewed, 13 unresolved discussions (waiting on @ahsanbarkati, @codelingo[bot], @manishrjain, and @vvbalaji-dgraph)

___
*[raftwal/log.go, line 321 at r2](https://reviewable.io/reviews/dgraph-io/dgraph/6714#-MJWJlM1M_8mWX-4hssi-r2-321:-MJX1mbFDzkyMP336HZC:b95vwze) ([raw file](https://github.com/dgraph-io/dgraph/blob/0454d5c69da86402a4dedfdf8c5cb892db1c37b2/raftwal/log.go#L321)):*
<details><summary><i>Previously, codelingo[bot] wrote…</i></summary><blockquote>

Prefer fixed size array allocations on stack over heap

[View Rule](https://dash.codelingo.io/repos/dgraph-io/dgraph/rules/1578)
</blockquote></details>

Could just directly point buf to the corresponding lf.Data.
___
*[raftwal/log.go, line 51 at r4](https://reviewable.io/reviews/dgraph-io/dgraph/6714#-MJX-lAlC4FH2iDYQobO:-MJX-lAmE8RUdHAPnkbl:b-4xm9ur) ([raw file](https://github.com/dgraph-io/dgraph/blob/7de21dce8876202158ac3571ae6809e6b8f35860/raftwal/log.go#L51)):*
> ```Go
> 	// logFileOffset is offset in the log file where data is stored.
> 	logFileOffset = 1 << 20 // 1MB
> 	// encryptionKeyOffset is offset in the log file where encryption key is stored.
> ```

The first baseIVsize bytes are for base IV and the rest of the bytes are for key ID.
___
*[raftwal/log.go, line 113 at r4](https://reviewable.io/reviews/dgraph-io/dgraph/6714#-MJX08Bp6Gz6qP2hN-Jk:-MJX08Bp6Gz6qP2hN-Jl:b6lbini) ([raw file](https://github.com/dgraph-io/dgraph/blob/7de21dce8876202158ac3571ae6809e6b8f35860/raftwal/log.go#L113)):*
> ```Go
> 			InMemory:                      false,
> 		}
> 		if lf.registry, err = badger.OpenKeyRegistry(krOpt); err != nil {
> ```

This won't open Badger. It would only use its key registry.
___
*[raftwal/log.go, line 120 at r4](https://reviewable.io/reviews/dgraph-io/dgraph/6714#-MJX2xzkC_cXf-NWMRa4:-MJX2xzkC_cXf-NWMRa5:b-92klli) ([raw file](https://github.com/dgraph-io/dgraph/blob/7de21dce8876202158ac3571ae6809e6b8f35860/raftwal/log.go#L120)):*
> ```Go
> 	fpath := logFname(dir, fid)
> 	// Open the file in read-write mode and create it if it doesn't exist yet.
> 	lf.MmapFile, err = z.OpenMmapFile(fpath, os.O_RDWR|os.O_CREATE, logFileSize)
> ```

If I have an existing file, which isn't encrypted, how do I know? Can I read that file?

If the key ID is zero, then we know.
___
*[raftwal/log.go, line 132 at r4](https://reviewable.io/reviews/dgraph-io/dgraph/6714#-MJX0LqD5fv9ACQbh5qM:-MJX0LqD5fv9ACQbh5qN:b-7ilvfu) ([raw file](https://github.com/dgraph-io/dgraph/blob/7de21dce8876202158ac3571ae6809e6b8f35860/raftwal/log.go#L132)):*
> ```Go
> 	} else {
> 		if len(encKey) != 0 {
> 			// Openend an existing file.
> ```

typo?
___
*[raftwal/log.go, line 134 at r4](https://reviewable.io/reviews/dgraph-io/dgraph/6714#-MJX1LEu210EEAJwW5Qz:-MJX1LEv0IM8Z-eIgQrv:b-kf2r8q) ([raw file](https://github.com/dgraph-io/dgraph/blob/7de21dce8876202158ac3571ae6809e6b8f35860/raftwal/log.go#L134)):*
> ```Go
> 			// Openend an existing file.
> 			buf := make([]byte, 16)
> 			copy(buf, lf.Data[encryptionKeyOffset:])
> ```

`buf := lf.Data[encryptionKeyoffset:+16]`
___
*[raftwal/log.go, line 178 at r4](https://reviewable.io/reviews/dgraph-io/dgraph/6714#-MJX1_ZF6-4n1aMHnaEa:-MJX1_ZF6-4n1aMHnaEb:b-dtwu1w) ([raw file](https://github.com/dgraph-io/dgraph/blob/7de21dce8876202158ac3571ae6809e6b8f35860/raftwal/log.go#L178)):*
> ```Go
> 	// Decrypt the data if encryption is enabled.
> 	if x.WorkerConfig.EncryptionKey != nil && len(re.Data) > 0 {
> 		// No need to worry about mmap. because, XORBlock allocates a byte array to do the
> ```

s/because/Because
___
*[raftwal/log.go, line 179 at r4](https://reviewable.io/reviews/dgraph-io/dgraph/6714#-MJX1dK8CYMhmFymljv1:-MJX1dK8CYMhmFymljv2:b2nhrme) ([raw file](https://github.com/dgraph-io/dgraph/blob/7de21dce8876202158ac3571ae6809e6b8f35860/raftwal/log.go#L179)):*
> ```Go
> 	if x.WorkerConfig.EncryptionKey != nil && len(re.Data) > 0 {
> 		// No need to worry about mmap. because, XORBlock allocates a byte array to do the
> 		// xor. So, the given slice is not being mutated.
> ```

XOR
___
*[raftwal/log.go, line 333 at r4](https://reviewable.io/reviews/dgraph-io/dgraph/6714#-MJX1s5S344Ed0k_yiST:-MJX1s5S344Ed0k_yiSU:byrvdoo) ([raw file](https://github.com/dgraph-io/dgraph/blob/7de21dce8876202158ac3571ae6809e6b8f35860/raftwal/log.go#L333)):*
> ```Go
> 	// Initialize base IV.
> 	lf.baseIV = buf[8:]
> 	copy(lf.Data[encryptionKeyOffset:], buf)
> ```

No need to copy after.
___
*[raftwal/wal.go, line 177 at r4](https://reviewable.io/reviews/dgraph-io/dgraph/6714#-MJX22eSDnaEB8JUr19q:-MJX22eSDnaEB8JUr19r:b9eflz4) ([raw file](https://github.com/dgraph-io/dgraph/blob/7de21dce8876202158ac3571ae6809e6b8f35860/raftwal/wal.go#L177)):*
> ```Go
> 
> 		// If encryption is enabled then encrypt the data.
> 		if x.WorkerConfig.EncryptionKey != nil {
> ```

If there's a valid curr.dataKey, then we should encrypt. Because, you're going to read that file later based on whether it has the key or not.

**Review by manishrjain — APPROVED** [review]
:lgtm:

> Reviewed 2 of 6 files at r1, 1 of 4 files at r2, 3 of 3 files at r5.
> *[Reviewable](https://reviewable.io/reviews/dgraph-io/dgraph/6714#-:-MJbQp0tDwFlgk5LpbXh:bnfp4nl)* status: all files reviewed, 16 unresolved discussions (waiting on @ahsanbarkati, @codelingo[bot], and @vvbalaji-dgraph)

___
*[raftwal/log.go, line 132 at r4](https://reviewable.io/reviews/dgraph-io/dgraph/6714#-MJX0LqD5fv9ACQbh5qM:-MJbPUWRCzsUiUuOnTei:bzi7xv) ([raw file](https://github.com/dgraph-io/dgraph/blob/7de21dce8876202158ac3571ae6809e6b8f35860/raftwal/log.go#L132)):*
<details><summary><i>Previously, manishrjain (Manish R Jain) wrote…</i></summary><blockquote>

typo?
</blockquote></details>

[encOffset: encOffset + 16]
___
*[raftwal/log.go, line 55 at r5](https://reviewable.io/reviews/dgraph-io/dgraph/6714#-MJbP2VM9aSM24qHLEkd:-MJbP2VNCkceu1nttGph:b-4lhfc0) ([raw file](https://github.com/dgraph-io/dgraph/blob/9d503536e96b3366b80864be58e27eea331a2c96/raftwal/log.go#L55)):*
> ```Go
> 	baseIVsize = 8
> 	// keyIDsize is the size of keyID.
> 	keyIDsize = 8
> ```

Just remove keyIDsize.
___
*[raftwal/log.go, line 138 at r5](https://reviewable.io/reviews/dgraph-io/dgraph/6714#-MJbPil714NQWZ8HcUnp:-MJbPil714NQWZ8HcUnq:b-vq8vta) ([raw file](https://github.com/dgraph-io/dgraph/blob/9d503536e96b3366b80864be58e27eea331a2c96/raftwal/log.go#L138)):*
> ```Go
> 	} else {
> 		buf := lf.Data[encryptionKeyOffset : encryptionKeyOffset+baseIVsize+keyIDsize]
> 		keyID := binary.BigEndian.Uint64(buf[:keyIDsize])
> ```

buf[:8]
___
*[raftwal/log.go, line 150 at r5](https://reviewable.io/reviews/dgraph-io/dgraph/6714#-MJbPmaK2eZtMWrLfd86:-MJbPmaK2eZtMWrLfd87:b-vq8x8y) ([raw file](https://github.com/dgraph-io/dgraph/blob/9d503536e96b3366b80864be58e27eea331a2c96/raftwal/log.go#L150)):*
> ```Go
> 				return nil, err
> 			}
> 			lf.baseIV = buf[keyIDsize:]
> ```

buf[8:]
___
*[raftwal/log.go, line 189 at r5](https://reviewable.io/reviews/dgraph-io/dgraph/6714#-MJbQ0ABACKHUICs44Kg:-MJbQ0ABACKHUICs44Kh:b-9vgsnb) ([raw file](https://github.com/dgraph-io/dgraph/blob/9d503536e96b3366b80864be58e27eea331a2c96/raftwal/log.go#L189)):*
> ```Go
> 		decoded, err := y.XORBlockAllocate(
> 			re.Data, lf.dataKey.Data, lf.generateIV(entry.DataOffset()))
> 		x.Check(err)
> ```

Just a note saying that we could potentially put these on an allocator.

**Pull request merged**